compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -16534,25 +16715,123 @@ TR::Node *nullchkSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier *
+      {
+      if (childOpCode.isDiv() || childOpCode.isRem())
+         {
+         s->_nodeToDivchk = NULL;
+
+         child = s->simplify(child, block);
 
-   if (child != originalChild ||
-       !(child->getOpCode().isDiv() || child->getOpCode().isRem()))
+         // Handshake with division and remainder simplifiers:  If simplifying the
+         // child of a DIVCHK results in a node that still needs to have a DIVCHK
+         // applied, those simplififiers will place that node in _nodeToDivchk.  If
+         // the simplification leaves no node that needs to have a DIVCHK applied,
+         // they will set _nodeToDivchk to NULL.  If _nodeToDivchk is non-null, that
+         // is used node as the child of the DIVCHK; otherwise, the DIVCHK is
+         // simplified to a treetop node.
+         //
+         // Note that the division and remainder simplifiers will set _nodeToDivchk
+         // only after recursively simplifying their own children, so any setting of
+         // _nodeToDivchk by division or remainder operations that are deeper in the
+         // trees will not be seen upon returning to this method.
+         //
+         // For example, before simplifying the DIVCHK child, n98n, we might have the
+         // trees on the left and after, the trees on the right.  Notice that the
+         // ldiv child of the DIVCHK has been replaced with an idiv that must still
+         // be checked for division by zero.  The ldivSimplifier will set
+         // _nodeToDivchk to refer to node n100n so that divchkSimplifier
+         // can update the DIVCHK node to refer to the correct child node.
+         //
+         // n99n  DIVCHK                    n99n  DIVCHK
+         // n98n    ldiv                    n100n   idiv
+         // n97n      i2l
+         // n96n        iload a             n96n      iload a
+         // n95n      i2l
+         // n94n        iload b             n94n      iload b
+         // n93n  lstore c                  n93n  lstore c
+         // n98n    ==> ldiv                n98n    i2l
+         //                                 n100n     ==> idiv
+         //
+         // If the numerator of a division is itself a division operation and the
+         // denominator is unity (1), the trees might look like those on the left,
+         // while if the numerator was simply the value of a variable, it might
+         // look like those on the right:
+         //
+         // n199n DIVCHK                       n299n DIVCHK
+         // n198n   idiv                       n298n   idiv
+         // n197n     idiv                     n297n     iload a
+         // n196n       iload a                n296n     iconst 1
+         // n195n       iload b
+         // n194n     iconst 1
+         //
+         // For the trees on the left, the result of simplifying the child of the
+         // DIVCHK would be n197n, which is itself an idev, while for the trees on
```

Comment:
`idev` -> `idiv`

Reply:
Fixed in commit 37c05e256709f96c1913b8fcab1b45b3e3b2dca2